WordPress/gutenberg
# ToggleGroupControl: Refactor to handle de-selection and multiple selections

Follow-up to #44067
Fixes icon color issue noted in https://github.com/WordPress/gutenberg/pull/44065#issuecomment-1245068026

I'll probably split this for code review.

🚩 In this PR, let's first discuss the visual/interaction design concerns.

## What?

The goal of this refactor is to add de-selection and multi-selection support to ToggleGroupControl, in a way that is:

- Sufficiently accessible through assistive technologies
- Visually clear and consistent
- Has non-confusing API for developers

## Why?

Discussions started in #36735 and many possible approaches were thought through. Eventually, we arrived at the point where product requirements pretty much necessitated a "de-selectable radio control", which initially seemed kind of impossible to implement accessibly. Our impasse was broken when @noisysocks found [some recent information](https://lea.verou.me/2022/07/button-group/) from two accessibility experts regarding almost our exact problem. According to the article, it's not only ok but *preferable* that segmented controls are implemented with `button` and `aria-pressed`.

So with the semantics issue unblocked, all we have to do is work out the visual design and API.

## How?

To recap, our component needs to support three combinations of behaviors:

1. mutually exclusive, not de-selectable
1. mutually exclusive, de-selectable
1. not mutually exclusive (de-selectable by nature)

### Mutually exclusive, not de-selectable

First, this is our basic "radio-like" behavior. This is how ToggleGroupControl works today. The entire control is one tab stop, and options are toggled immediately when navigating with the arrow keys.

https://user-images.githubusercontent.com/555336/190493709-0a43e402-6c1f-442a-a087-3467967d6e6e.mp4

### Mutually exclusive, de-selectable

Next, we have a mutually exclusive, but de-selectable behavior. The button options are now distinct, and separate tab stops.

https://user-images.githubusercontent.com/555336/190493101-ef38239e-59c6-474a-89de-73b53bbe48f8.mp4

There are actually already a good number of ad hoc components in Gutenberg that pretty much look like ☝️, sans backdrop animation. However, some initial feedback from the Components Squad noted that it was visually unclear that this was a group of buttons, and was suggested that maybe we should keep the border. After trying it, I think I prefer the borderless version, but here's a bordered one for reference:

https://user-images.githubusercontent.com/555336/190493108-7e8074db-b345-4f93-a51e-4cdfdb13ded0.mp4

### Not mutually exclusive (always de-selectable)

Finally, this is the multi-select version. As far as I know, the only similar UI that currently exists in Gutenberg is the formatting buttons in the Rich Text toolbar:

<img width="159" alt="Bold and Italic buttons in the inline block toolbar" src="https://user-images.githubusercontent.com/555336/190496500-fc995131-957f-450a-9533-796409c5a234.png">

https://user-images.githubusercontent.com/555336/190493099-72a1ce2f-a887-4a7d-a3f4-3a06440b1556.mp4

(☝️ Sorry the icons are a smidge off-center. I'll fix that)

Here there was also some initial feedback from the Components Squad that the "group"-ness was visually unclear. A version with some kind of separator was suggested, so here's a rough prototype for reference:

https://user-images.githubusercontent.com/555336/190493080-c9ca0282-2dfb-4656-8410-eb59ed25d585.mp4


## Testing Instructions

The docs are incomplete, but you can do `npm run storybook:dev` and see the stories for ToggleGroupControl and ToggleMultipleGroupControl if you want to try it out yourself.
